### ISSUE TITLE
OpenMSX won't boot

### ISSUE THREAD

**Issue by Colpocorto** [openMSX/openMSX#1500]
OpenMSX crashes during startup.

Version: OpenMSX 18.0 (2022-06-12 release).
OS: Windows 10 x64
Guest machines: failure reproduced while emulating CBIOS MSX2+, Canon V-20, Panasonic FS-A1GT, among others.
Hardware: HP Probook 450 G8

Tested with regular Windows user and administrator user, same results.

Two events are logged on the Windows Event Viewer (view attached files).
[OpenMSX_err.zip](https://github.com/openMSX/openMSX/files/11240386/OpenMSX_err.zip)

**Comment by m9710797** [comment]
Hi,

I'd like to help, but I've no idea how to view .evtx files. (I don't have access to a windows machine.)
Can you attach the same info in readable form?

Wouter

**Comment by MBilderbeek** [comment]
So, did it work before and stopped working? I think you have been using openMSX before, right? Or is this the first time you try it on this computer?

**Comment by Colpocorto** [comment]
> So, did it work before and stopped working? I think you have been using openMSX before, right? Or is this the first time you try it on this computer?

It worked before on this machine, but the whole OS has been reinstalled so it's a fresh installation of both Windows and OpenMSX.

**Comment by MBilderbeek** [comment]
And you are using the same openMSX version as before the reinstall?

Can you show what you're seeing? Popups? Log entries from event log?

**Comment by Colpocorto** [comment]
> Hi,
> 
> I'd like to help, but I've no idea how to view .evtx files. (I don't have access to a windows machine.) Can you attach the same info in readable form?
> 
> Wouter

Hi Wouter, let me figure out how to convert .evtx files to readable text (I've managed to convert them to XML at least).  In any case, how will you reproduce the issue without a Windows machine?

**Comment by Colpocorto** [comment]
> And you are using the same openMSX version as before the reinstall?
> 
> Can you show what you're seeing? Popups? Log entries from event log?

Hi Manuel,

thank you for your response. It is exactly the same version as before. The chronology was as follows:

-My company gave me a brand new laptop with Windows 10.
-I installed OpenMSX 18  (in fact, the same .MSI file)
-I upgraded the OS to Windows 11. OpenMSX still worked perfectly.
-I experienced tons of W11 related issues (slowness, excessive consumption of RAM, etc.) so I asked the IT guys to downgrade the laptop to Windows 10 again. This operation included a full SDD formatting and installing W10 from the original image.
-I installed OpenMSX on the new W10 but it does not work anymore.

No logs or popups screens. The windows frame appears (with no graphics inside), but then the windows dissapears entirely without showing any error. The only trace of the problem is shown on the Windows Event Viewer.

@m9710797 Wouter, this is the text contents of the Windows events:

Event 1:
----------------------------------------------------------------------------------
Faulting application name: openmsx.exe, version: 18.0.0.0, time stamp: 0x62a6138a
Faulting module name: unknown, version: 0.0.0.0, time stamp: 0x00000000
Exception code: 0xc0000005
Fault offset: 0x0000000000000000
Faulting process ID: 0x3c60
Faulting application start time: 0x01d96fc80b1692a1
Faulting application path: C:\Program Files\openMSX\openmsx.exe
Faulting module path: unknown
Report ID: 89482b8c-8efe-4910-8a49-e52ea8f82408
Faulting package full name: 
Faulting package-relative application ID: 
------------------------------------------------------------------------------

Event 2:
----------------------------------------------------------------------------------
Cab Id: 0

Problem signature:
P1: openmsx.exe
P2: 18.0.0.0
P3: 62a6138a
P4: StackHash_ac46
P5: 0.0.0.0
P6: 00000000
P7: PCH_93_FROM_ntdll+0x000000000009DBB4
P8: c0000005
P9: 0000000000000008
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER24C7.tmp.dmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER264F.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER266F.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER267C.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER2777.tmp.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_openmsx.exe_28681a9da70c63af2a555b825c4bf6c9f29af9a_9ca5b21f_c70e3387-0836-4903-9bbc-4874d45011df

Analysis symbol: 
Rechecking for solution: 0
Report Id: 2a9bcd05-ad66-403e-9e24-cf8509258858
Report Status: 268435456
Hashed bucket: c3759307a1b72c90ee1a5319a5a39953
Cab Guid: 0
----------------------------------------------------------------------------

By the way, the "attached files" referenced in the error message do not actually exists on that path.

**Comment by Colpocorto** [comment]
Additional info: I've used [ProcMon](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) to log all the activity generated by the openmsx.exe process:

[Logfile.CSV](https://github.com/openMSX/openMSX/files/11242271/Logfile.CSV)

**Comment by MBilderbeek** [comment]
OK, I have no clue what could be going on. But it seems something on your laptop has triggered it. Is it possible that software on the laptop is somehow blocking some part of openMSX, which causes it to crash? 
I suppose you could try to run openMSX with a debugger (and use the pdb files with it) to see where it crashes. There's no way we can reproduce this. (Other people would have reported it as well, I guess...)

**Comment by Colpocorto** [comment]
> OK, I have no clue what could be going on. But it seems something on your laptop has triggered it. Is it possible that software on the laptop is somehow blocking some part of openMSX, which causes it to crash? I suppose you could try to run openMSX with a debugger (and use the pdb files with it) to see where it crashes. There's no way we can reproduce this. (Other people would have reported it as well, I guess...)

Agree, it definitely makes sense.

I guess that the culprit might be the corporate antivirus. A bit strange, though, since it is exactly the same antivirus (with the same configuration) that the laptop had in the past and it never caused trouble.  In fact, the whole PC configuration it's image-driven, so it should be absolutely identical (except for the hostname and the IP address) to the previous installation.

I'll try debugging the openmsx.exe process with the aid of the pdb files and see what's going on. Even if it hasn't been reported by others yet, knowing the cause of the problem might be useful in the future.

Thank you for your help.

**Comment by MBilderbeek** [comment]
OK, looking forward to the results!

**Comment by Colpocorto** [comment]
Here it is:

![image](https://user-images.githubusercontent.com/24856281/233992273-6fe2193e-aed3-41f9-8f4a-53074cb814bb.png)

It seems to be the Glew library...

**Comment by Colpocorto** [comment]
I've been able to reproduce the problem with both the officially released "openmsx.exe" executable file and the one built by myself on VS2022.  There is something in my laptop that makes the glCreateProgram function to fail (and it seems that the exception is not caught properly).

Edit: OpenMSX starts correctly if the renderer is set to SDL instead of SDLGL-PP.

**Comment by MBilderbeek** [comment]
You could try to upgrade the glew library to the latest version... But
there isn't much else we can do about, I'm afraid. Perhaps upgrade your
graphics drivers...?

Op ma 24 apr. 2023 14:43 schreef Colpocorto ***@***.***>:

**Comment by Colpocorto** [comment]
I can try updating the drivers, although the same drivers worked in the past.

In any case, the emulator crashes in a bad way, it looks a bit amateurish IMHO. Adding a better exception control and error display would be a good thing. I think there is an improvement opportunity here.

**Comment by MBilderbeek** [comment]
How would you propose to do that, as it crashes inside the call of the glew
library?
Patch is most welcome.

Op di 25 apr. 2023 09:45 schreef Colpocorto ***@***.***>:

**Comment by m9710797** [comment]
Hi Colpocorto,

Thanks for the extra info!
I agree with you (in principle): better error handling would be nice. Though at the moment it's not yet clear how to achieve that.
From the screenshot you posted it seems that either:
* openMSX crashes somewhere inside the `glCreateProgram()` function.
* `glCreateProgram` is a function pointer which is a nullptr.

If it's the former, then I'm not sure what we can do about it.

If it's the latter, then maybe we're missing some error check earlier in openMSX? Just a wild guess: maybe your system only offers openGL version 1.x? The function `glCreateProgram()` is only available in openGL version 2.x, and then maybe glew sets this function pointer to nullptr?

I would really like to fix/improve this, because in (one of) the next openMSX releases we'll drop the SDL renderer. In other words: we will require openGL version 2.x.

@Colpocorto we need your help to continue to investigate this. I hope you're able/willing to do this. Thanks.

Wouter

**Comment by Colpocorto** [comment]
> How would you propose to do that, as it crashes inside the call of the glew library? Patch is most welcome. 

Patch the library and add some exception handling to glCreateProgram.

**Comment by Colpocorto** [comment]
> Hi Colpocorto,
> 
> Thanks for the extra info! I agree with you (in principle): better error handling would be nice. Though at the moment it's not yet clear how to achieve that. From the screenshot you posted it seems that either:
> 
> * openMSX crashes somewhere inside the `glCreateProgram()` function.
> * `glCreateProgram` is a function pointer which is a nullptr.
> 
> If it's the former, then I'm not sure what we can do about it.
> 
> If it's the latter, then maybe we're missing some error check earlier in openMSX? Just a wild guess: maybe your system only offers openGL version 1.x? The function `glCreateProgram()` is only available in openGL version 2.x, and then maybe glew sets this function pointer to nullptr?

In fact, it seems that ALL the functions have assigned a null pointer:

![image](https://user-images.githubusercontent.com/24856281/234257343-e8c3c465-6619-4c3c-b8f0-bdf447dfa5e2.png)
![image](https://user-images.githubusercontent.com/24856281/234257426-ef3e708a-8313-4365-b212-a21626875710.png)


> 
> I would really like to fix/improve this, because in (one of) the next openMSX releases we'll drop the SDL renderer. In other words: we will require openGL version 2.x.
> 
> @Colpocorto we need your help to continue to investigate this. I hope you're able/willing to do this. Thanks.
> 

Your guess is correct. My laptop is OpenGL 1.1 (a bit strange, since it has been installed from the very same image used by IT crew in the past).

![image](https://user-images.githubusercontent.com/24856281/234258054-a75968ea-4289-4d22-8a03-4268a15db023.png)

If you finally drop SDL and only support GL, the user should be warned when trying to open OpenMSX.exe instead of creashing without any kind of handling.  Maybe just query the installed OpenGL version and quit with a message popup if the system is not supported.

**Comment by Colpocorto** [comment]
UPDATE: the problem with my laptop has been solved by updating the video drivers. Now the system is OpenGL v4.6 and OpenMSX works perfectly.

Bottomline: the issue was caused by an unsupported OpenGL version and, although it is not a bug itself, a warning/notification message instead of crashing would be a good improvement.

Thank you for your advice, guys.

**Comment by m9710797** [comment]
I fully agree we should have a clear error message when the openGL version is insufficient.
And till half an hour ago I thought we had such a check:
* src/video/SDLGLVisibleSurface.cc, line 42: we request a openGL 2.1 context.
* src/video/SDLGLVisibleSurface.cc, line 56: we report an error when this context couldn't be created.

Unfortunately the SDL documentaton isn't very clear on this topic. When re-reading the documentaton, I **think** `SDL_GL_CreateContext()` may also create a context with a **lower** version number than requested. If true, that means you still have to explicitly check the obtained version afterwards. I'll try to add this extra check.

I'm not sure how I can test this? Is there a way to (temporary) downgrade the available openGL version on a (linux) system?

Thanks for your help!

**Comment by Colpocorto** [comment]
Maybe on a Windows 7 virutalbox machine and the minimum set of drivers available?

**Comment by MBilderbeek** [comment]
Can we somehow check whether these functions are a nullptr and use that as a check?

**Comment by m9710797** [comment]
I did some digging through our commit history:
This commit broke the openGL version check: c3679faad640f9b8d82f59aeef51294ba0dc517c
(Even though the change looked totally reasonable when I reviewed it).

We should re-add a check like `if (!GLEW_VERSION_2_1) { ...`, but I didn't have time yet to do that.

**Comment by Colpocorto** [comment]
> I would really like to fix/improve this, because in (one of) the next openMSX releases we'll drop the SDL renderer. In other words: we will require openGL version 2.x.
> Wouter

One question about this: how will it work on Mac? OpenGL support was deprecated on Mac OS since Catalina (and that's 4 years old now).

**Comment by m9710797** [comment]
Once OpenGL is no longer supported on Mac, we'll need a volunteer to port our OpenGL-specific code to Metal. Or maybe Vulkan if that's available on Mac? (Or possibly there are, or will be, libraries that translate OpenGL calls into Metal or Vulkan?)

The reason for this change (dropping software rendering) is that we're working on integrating a GUI inside openMSX itself. We've picked ["Dear ImGui"](https://github.com/ocornut/imgui) as our cross-platform toolkit. And that one really needs some form of hardware-accelerated graphics API (can be OpenGL, DirectX, Metal, Vulkan, ...). We've picked OpenGL because that's cross-platform and much easier to use than Vulkan (and sufficient for our needs).
